compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
+  Diagnostics stage ->
+  IO (Diagnostics stage)
+setStageDiagnostics fp timeM stage diags (Diagnostics ds) = do
+    uri <- toUri fp
+    let thisFile = Map.lookup uri ds
+        addedStage :: StoreItem
+        addedStage = StoreItem (Just posixTime) $ storeItem thisFile
+    pure $ Diagnostics $ Map.insert uri addedStage ds where
+
+    storeItem :: Maybe StoreItem -> DiagnosticsBySource
+    storeItem = \case
+        Nothing ->
+            Map.singleton k v
+
+        Just (StoreItem Nothing _) ->
+            noVersions
+
+        Just (StoreItem (Just prevTime) bySource) ->
+            case compare posixTime prevTime of
+                EQ -> insert bySource
+                GT -> insert $ Map.map (SL.map tagOutdated)  bySource
+                LT -> bySource
+
+    -- Change the color of outdated diagnostics
+    tagOutdated :: Diagnostic -> Diagnostic
+    tagOutdated d = d{_severity = Just DsInfo}
+
+    isOutdated :: Diagnostic -> Bool
+    isOutdated Diagnostic{_severity} = _severity == Just DsInfo
+
+    removeOutdated, insert :: DiagnosticsBySource -> DiagnosticsBySource
+    removeOutdated = Map.filter (not . any isOutdated)
+
+    insert bySource =
+        let removeExpired =
+                if Map.lookup k bySource == Just (SL.map tagOutdated v)
+                then removeOutdated
+                else id
+        in Map.insert k v $ removeExpired bySource
+
+    k = Just $ T.pack $ show stage
+    v = SL.toSortedList diags
+
+    posixTime :: Int
+    posixTime = maybe 0 (fromEnum . utcTimeToPOSIXSeconds) timeM
+
+noVersions :: HasCallStack => a
+noVersions =
+    error "All store items must have versions"
+
+fromUri :: LSP.Uri -> FilePath
+fromUri = fromMaybe noFilePath . uriToFilePath
+
+-- TODO (MK) We canonicalize to make sure that the two files agree on use of
+-- / and \ and other shenanigans.
+-- haskell-lsp also has serious problems with relative file paths
+toUri :: FilePath -> IO LSP.Uri
+toUri = fmap filePathToUri . canonicalizePath
+
+getAllDiagnostics ::
+    Diagnostics stage ->
+    [FileDiagnostic]
+getAllDiagnostics =
+    concatMap (\(k,v) -> map (fromUri k,) $ getDiagnosticsFromStore v) . Map.toList . getStore
+
+getFileDiagnostics ::
+    FilePath ->
+    Diagnostics stage ->
+    IO [LSP.Diagnostic]
+getFileDiagnostics fp ds = do
+    uri <- toUri fp
+    pure $
+        maybe [] getDiagnosticsFromStore $
+        Map.lookup uri $
+        getStore ds
+
+getStageDiagnostics ::
+    Show stage =>
+    FilePath ->
+    stage ->
+    Diagnostics stage ->
+    IO [LSP.Diagnostic]
+getStageDiagnostics fp stage (Diagnostics ds) = do
+    uri <- toUri fp
+    pure $ fromMaybe [] $ do
+        (StoreItem _ f) <- Map.lookup uri ds
+        toList <$> Map.lookup (Just $ T.pack $ show stage) f
+
+filterDiagnostics ::
+    (FilePath -> Bool) ->
+    Diagnostics stage ->
+    Diagnostics stage
+filterDiagnostics keep =
+    Diagnostics .
+    Map.filterWithKey (\file _ -> maybe False keep $ uriToFilePath file) .
```

Comment:
Under which circumstances do we expect `uriToFilePath` to fail? When we use `noFilePath`? If so, do  we really want to filter these?